llvm/include/llvm/CodeGen/SDPatternMatch.h

```diff
@@ -457,6 +457,53 @@ struct BinaryOpc_match {
   }
 };
 
+template <typename LHS_t, typename RHS_t, typename PredFuncT>
+struct AnyBinaryOp_match {
+  LHS_t L;
+  RHS_t R;
+  PredFuncT PredFunc;
+  bool Commutable;
+
+  AnyBinaryOp_match(const PredFuncT &Pred, const LHS_t &LHS, const RHS_t &RHS,
+                    const bool Commutable)
+      : PredFunc(Pred), L(LHS), R(RHS), Commutable(Commutable) {}
+
+  template <typename OpTy, typename MatchContext>
+  bool match(OpTy *V, const MatchContext &Ctx) {
+    assert(Ctx.getTLI() && "TargetLowering is required for this pattern");
+    if (auto *I = dyn_cast<BinaryOperator>(V))
+      return (PredFunc(*Ctx.getTLI()) &&
+              ((L.match(I->getOperand(0)) && R.match(I->getOperand(1))) ||
+               (Commutable && L.match(I->getOperand(1)) &&
+                R.match(I->getOperand(0)))));
+    return false;
+  }
+};
+
+template <typename LHS, typename RHS>
+inline auto m_AnyBinOp(const LHS &L, const RHS &R) {
+  return AnyBinaryOp_match{[](const TargetLowering &TLI) { return true; }, L, R,
```

Comment:
I think there is a misunderstanding here: the reason m_AnyBinOp is called "any" is because, as opposed to the existing m_BinOp who matches against a _fixed_ opcode given by user, m_AnyBinOp will match _any_ opcode that is TLI.isBinary/isCommutativeBinOp without any user-provided opcode to compare against with so you couldn't just return true here. 
Furthermore, `m_AnyBinOp(unsigned &Opc, LHS, RHS)` is basically extracting the opcode that fulfills `m_AnyBinOp(LHS, RHS)`. Therefore, I don't think PredFunc is necessary since both variants of `m_AnyBinOp` has to call TLI.isBinary/isCommutativeBinOp.

Reply:
Can you elaborate what you mean by: 
> Furthermore, m_AnyBinOp(unsigned &Opc, LHS, RHS) is basically extracting the opcode that fulfills m_AnyBinOp(LHS, RHS).

Reply:
> Can you elaborate what you mean by:
> 
> > Furthermore, m_AnyBinOp(unsigned &Opc, LHS, RHS) is basically extracting the opcode that fulfills m_AnyBinOp(LHS, RHS).

So `m_AnyBinOp(LHS, RHS)` would check if the SDNode's opcode fulfills TLI.isBinOP/isCommutativeBinOp; `m_AnyBinOp(unsigned &Opc, LHS, RHS)` is basically doing the same thing `m_AnyBinOp(LHS, RHS)` does, plus returning the opcode (that has been checked) via the `Opc` argument (as `Opc` has a type of unsigned reference, making it an output argument).

Reply:
I have updated the impl. Please let me know if that is what was required.

Reply:
Okay, so while I think that the ```m_AnyBinOp(unsigned &Opc, LHS, RHS)``` needs some work in the sense that I am still not returning the opcode, ie. giving Opcode the value of ```N->getOpcode()```. But according to my testing, I don't think that ```isBinOp``` is being called at all. Any reason why?

Reply:
> I don't think that isBinOp is being called at all

You could attach a debugger to see why isBinOp is skipped